When GlobalJitLevel is on, disable the Grappler memory opt.

trentlo opened the pull request:
This commit disables the Grappler memory optimizer when the XLA JIT is detected on.

The (current) XLA clustering can result in loss of concurrency between kernel compute and memory copies. As such, it usually loses the concurrency needed to hide the latencies of the inserted swap-ins and swap-outs and incurs great performance overhead.

You may find more details about the performance degradation in this document:
https://docs.google.com/document/d/1q1UPN2CRRNBoUXM0zORT-cTG9m7ctQRB2xfUPhaK1Ek/edit?usp=sharing

@sanjoy, please also help to take a look.

trentlo committed:
Disable the Grappler memory opt when GlobalJitLevel is on.

trentlo committed:
Check GlobalJitLevel only for DEFAULT_MEM_OPT.

This relaxes the disable check and should be a slightly better behavior, as
users still have some ways to enable the memory optimizer when they want to.

trentlo commented:
One question:

According to `GetGlobalJitLevelForGraph()`, there are two ways to enable the XLA JIT; one is through the `ConfigProto` and the other is through the environment variable.

This PR can only check the ConfigProto and cannot check the env var flags as they only exist in the XLA world.

Do you think this is an issue? Any suggestions to address it?

nouiz commented:
maybe call getenv and parse the env variable?

sanjoy commented:
> This PR can only check the ConfigProto and cannot check the env var flags as they only exist in the XLA world.

I think we should just expose `GetGlobalJitLevelForGraph` to to TensorFlow by moving it to a header/cc outside compiler/xla.

@rmlarsen WDYT?

trentlo commented:
> > This PR can only check the ConfigProto and cannot check the env var flags as they only exist in the XLA world.
> 
> I think we should just expose `GetGlobalJitLevelForGraph` to to TensorFlow by moving it to a header/cc outside compiler/xla.
> 
> @rmlarsen WDYT?

It is more than that, as the `GetGlobalJitLevelForGraph` depends on compiler/jit/flags.cc, etc. which do not exist in a Tensorflow build without enabling XLA.

If we want to move the function, we will have to move several xla flag related files with it. Considering that, below are two options we have.

1. Just check the proto config without checking the env variables. This may make sense as the proto config should be the way for clients to enable XLA. Env variables are more for testing? Let me know if this is not the case though.
2. Move the xla flags related files into TF.

Personally, I am slightly leaning towards option 1 but please give some guidance. Thanks.

trentlo commented:
> maybe call getenv and parse the env variable?

IMHO, re-implement the parsing logic in TF may be ugly because there will then be two separate places to be maintained for future env variable change.

sanjoy commented:
> This may make sense as the proto config should be the way for clients to enable XLA. Env variables are more for testing? Let me know if this is not the case though.

Env vars are less elegant but they're super convenient when trying out XLA ("no need to make code changes, just flip this flag") so I'd like to make this work with env vars if possible.

Another possibility is to do some form of dependency injection via a registry -- allow the XLA JIT to register a callback that lets TF query whether the XLA global jit is enabled.  This will be more code but should be cleaner overall.

trentlo commented:
> Another possibility is to do some form of dependency injection via a registry -- allow the XLA JIT to register a callback that lets TF query whether the XLA global jit is enabled. This will be more code but should be cleaner overall.

Good suggestion. Let's proceed with this route.

trentlo committed:
Add a XlaConfigProxy class.

Callbacks can be registered to this class so that runtime environment
flags can be parsed to change configs in the Tensorflow core. A primary use
of this is for the Tensorflow core to query the XLA JIT level, which can be
configured by some runtime environment flags in addition to ConfigProto.

trentlo committed:
Add a unittest for XlaConfigProxy.

trentlo committed:
Fix a typo in XlaConfigProxy.

trentlo commented:
I implemented a config proxy in the Tensorflow core per our previous discussion. Please help to take a look when you have a moment. Thanks.

sanjoy commented on tensorflow/core/util/xla_config_proxy_test.cc:
```diff
@@ -0,0 +1,61 @@
+
+  // Register callbacks.
+  XlaConfigProxy::ConfigSetterRegistration<ConfigA>
+  contrig_setter_registration_a([](ConfigA& config_a) -> bool {
```
`contrib_setter_registration_a`

trentlo replied:
Execute me for the typo. Will fix them immediately.

sanjoy commented on tensorflow/core/util/xla_config_proxy.h:
```diff
@@ -0,0 +1,85 @@
+namespace tensorflow {
+
+// A proxy class of XLA config.
+class XlaConfigProxy {
```
I'm not sure I understand how you're planning to use this, do you mind elaborating a bit?

Here is what I had in mind though:  have a registry that lets XLA register an `std::function<OptimizerOptions::GlobalJitLevel(const GraphOptimizationPassOptions& options)>`.  Then Grappler can check if XLA is enabled using this registered `std::function`; if nothing is registered then it can assume that XLA is not enabled.

trentlo replied:
Very similar. Just that we need a singleton (or at least some static functions) for registering this `std::function` callback. And I slightly generalize the registry class into `ConfigSetterRegistry`.

`ConfigSetterRegistry` is a singleton utility class to register a `std::function<bool(ConfigType&)>`; meaning that it allows each different config type to register a callback. In the current case, XLA registers a `std::function<bool(GlobalJitLevel&)>`, and TF can then invoke this `std::function` to update the GlobalJitLevel value.

Perhaps the generalization loses the readability. If that is the case, let me remove the `ConfigSetterRegistry` and simply make `XlaConfigProxy` a singleton. Please feel free to let me know what you think.

sanjoy replied:
I think this generality is not YAGNI for now.

trentlo replied:
I see your point.

trentlo reviewed (commented):
Thanks for the review. Response inlined.

trentlo committed:
Fix typo.

trentlo committed:
Simplify XlaConfigProxy by removing the utililies.

Also, rename XlaConfigProxy to XlaConfigRegistry.

trentlo commented on tensorflow/compiler/jit/xla_cluster_util.cc:
```diff
@@ -224,13 +225,9 @@ struct XlaGlobalJitLevel {
 };
 
 XlaGlobalJitLevel GetXlaGlobalJitLevel(
-    const GraphOptimizationPassOptions& options) {
```
I changed this interface a bit as `GraphOptimizationPassOptions` is not accessible in meta_optimizer.cc.

trentlo commented on tensorflow/compiler/jit/xla_cluster_util.cc:
```diff
@@ -386,4 +388,19 @@ XlaAutoClusteringSummary GetXlaAutoClusteringSummary(const Graph& graph) {
+    OptimizerOptions::GlobalJitLevel jit_level_in_session_opts) {
+  XlaGlobalJitLevel xla_global_jit_level =
+      GetXlaGlobalJitLevel(jit_level_in_session_opts);
+  // Take the general flag to avoid the dependency on Tensorflow::Graph.
```
I simply ignore the single_graph flags, as the graph is not accessible in `InitializeOptimizers()`, where grappler configs its optimizers.

It makes sense to me that we simply expose the general flag (,although both of the general and single_gpu flags are currently the same). Let me know what you think about this though.

sanjoy replied:
I'm worried that this makes `XlaConfigRegistry::GetGlobalJitLevel` "look" misleading.

How about this:  we make the callback return an instance of `XlaGlobalJitLevel`.  Then in optimizers/meta_optimizer.cc we do:

```
XlaGlobalJitLevel xla_global_jit_level = ...;
if (xla_global_jit_level.general is ON) {
  Disable memory opt
}
// We don't care about single GPU because the regression happens only on multi-GPU 
```

This is a bit of a hack because if the graph happens to be single GPU but the general auto-jit is enabled then we're still going to disable the optimization but I think for getting us started this is fine.  Behaviorally this is identical to what you have but I think it is clearer.

Also a bit more context: if only single_gpu is enabled we absolutely don't want any behavior change on multi-GPU graphs.  We're using `--tf_xla_auto_jit=single-gpu(2)` to stage the rollout of XLA and we want `--tf_xla_auto_jit=single-gpu(2)` to only affect single GPU graphs.

trentlo replied:
I see your point. Thanks for the good feedback.

I made some changes:
1. Made the callback return `XlaGlobalJitLevel`, as this way is indeed less ambiguous.
2. I added check in the `MetaOptimizer` such that the memory optimizer is turned off only when JIT is on for _both_ single-gpu and multi-gpu graphs. This is a conservative approach and should be good enough for now.

Thanks for the context about how the single-gpu flag is used. It was not clear to me and it looks good to me.

Note that the swap-in/out latency issue should also exist in single gpu since the copy stream is different from device compute stream in the default Tensorflow setting.

trentlo reviewed (commented):
Updated. Please help to take a look again. Thanks.

trentlo committed:
Minor revision of header macros in xla_config_registry.h.

trentlo committed:
Revise XlaConfigRegistry.

1. Return XlaGlobalJitLevel instead of GlobalJitLevel in GetGlobalJitLevel().
2. Turn off the memory optimizer only when JIT is on for both single-gpu and
   multi-gpu. This is a more conservative behavior.
3. Naming style conformance.

trentlo reviewed (commented):
Thanks for the prompt and helpful feedback. I made some more changes accordingly. Please help to take yet another look.

trentlo committed:
Minor comment polishing.

trentlo committed:
Address review comments about XlaConfigRegistry.

1. Fix an oversight that the macro does not get expanded.
2. Add a mutex to the Register() method.
3. Comment polishing.

trentlo reviewed (commented):
I made changes addressing your new round of review comments. Thank you.

trentlo committed:
Guard both getter and setter in XlaConfigRegistry.

trentlo committed:
Use `new` instead of static object to avoid destructor.

By avoiding destructor, it should work better with multi-threaded program.

trentlo reviewed (commented):
Changes are made based on the suggestion. Thanks.

trentlo committed:
Add GUARDED_BY for GlobalJitLevelState.

trentlo reviewed (commented):
Revised. Thanks.

rthadur commented:
@trentlo can you please resolve conflicts ?

trentlo committed:
Merge branch 'official_master' into no_mem_opt_if_jit_on

trentlo commented:
> @trentlo can you please resolve conflicts ?

Yes, done.

trentlo commented:
@sanjoy 

Thanks for your time guiding me through this PR. I feel that reviewing this PR ends up being a tedious process.

trentlo commented:
@rthadur 

I noticed that the `Ubuntu CC` test may be stuck. Will that block the merging?

sanjoy commented:
> Will that block the merging?

I'm taking care of the merge.  Sorry for the delay, I'm a bit busier than usual.

trentlo commented:
> I'm taking care of the merge. Sorry for the delay, I'm a bit busier than usual.

Got it. No worries and take time then. Thanks.

byronyi commented:
@trentlo @goldiegadde I am not sure if this PR catch the r2.0 cherry-pick cycle. It does seem to fix a performance regression running the official ResNet50 model w/ XLA 8-GPU. 

I observed significant performance boost (~3900 image/s to ~7500 image/s) comparing 2.0 nightly packages in 2019-09-19 (wo/ this PR) and 2019-09-20 (w/ this PR).

Ping @tfboyd to confirm.

byronyi commented:
I will cherry-pick this on my own r2.0 branch and report the numbers tomorrow.

trentlo commented:
@byronyi FYI as below.

Indeed, cherry-pick the commits and measure the performance difference is the most direct way to verify.

A quick way to verify is to use the following flags. It dumps a bunch of logged files into `./graph_dump`.
`TF_DUMP_GRAPH_PREFIX="./graph_dump" TF_XLA_FLAGS="--tf_xla_clustering_debug=true" python3 ...`

Then you can use `grep -r GpuToHost ./graph_dump` to see whether any copy ops are inserted by Grappler. If you see them, likely they can cause great performance penalty. These ops should disappear when XLA is ON if this PR is there.

tfboyd commented:
I am seeing the same bounce back in the nightly tests.  XLA stated they were not focused on multi-GPU tests as their focus was XLA on by default for single GPU so I stopped bothering them with the results.  The drop to ~3,900 occurred around 03-AUG in my testing.  I see the bounce back to 8,500 with 2.0.0-dev20190922.  This is still not back to the ~9K we saw before the drop.  I doubt this will get cherry-picked into TF 2.0 as XLA is still experimental and RC2 is likely to be the last RC needed before release.

byronyi commented:
@tfboyd Bisection shows 53bdcf55ce180f609797746992987ce89c3543bb was the commit that causes bad performance in the nightly package, not the issue this PR fixes. So there must be some other changes that causes the performance degradation in r2.0. Do you mind to point to me a snapshot with the expected performance, i.e. before dropping to ~3,900 around 03-AUG?

tfboyd commented:
@byronyi   Here is the info I have that may be of the most use.  It would be nice to get back to the 9K, but it is also possible too many other changes have occurred. I only run on nightly and when nightly works.

Command for the test you would need a copy of imagenet.

```bash
# Flags
--batch_size=2048 --data_dir=/data/imagenet --dtype=fp16 --enable_eager --enable_xla --log_steps=10 --model_dir=/workspace/benchmarks/perfzero/workspace/output/2019-08-02-12-50-22-843844/benchmark_xla_8_gpu_fp16 --num_gpus=8 --noreport_accuracy_metrics --skip_eval --train_steps=110

# [PerfZero](https://github.com/tensorflow/benchmarks/tree/master/perfzero) command.  I suggest using perfzero when possible as it helps ensure your
# args are the same and a common docker is used for repeat-ability.  Not perfect but lightweight
# and improving.  
python3 /workspace/benchmarks/perfzero/lib/benchmark.py --benchmark_methods=official.benchmark.keras_imagenet_benchmark.Resnet50KerasBenchmarkReal.benchmark_xla_8_gpu_fp16 --data_downloads=gs://tf-perf-imagenet-uswest1/tensorflow/imagenet --python_path=models --git_repos="https://github.com/tensorflow/models.git;master;6b586a910d74a44f57da4d2335c79a20dc2803ab"
````

**GOOD**
07/31
8,723.9 to 9,000 images/sec
tensorflow hash:3e0ad8a004
model garden hash: 13e7c85d521d7bb7cba0bf7d743366f7708b9df7


**BAD**
08/03
We had a break in the builds for a few days and a big change to the default code path I believe.  Many problems
2,764.3 to 2,900 images/sec.
tensorflow hash: 3205135
model garden: 6b586a910d74a44f57da4d2335c79a20dc2803ab

**Current**
09/26
8,300 to 8,500 images/sec
tensorflow hash: 1ae01a0
model garden: 6f1e3b38d80f131e21f0721196df7cfc5ced2b74

It might be hard to bisect as I think the model garden code changed significantly between the two dates with the [following lines ](https://github.com/tensorflow/models/blob/master/official/vision/image_classification/resnet_imagenet_main.py#L168)and we also moved the code around recently so running the test now vs back in August there is a different directory:

```python
   # TODO(b/138957587): Remove when force_v2_in_keras_compile is on longer
    # a valid arg for this model. Also remove as a valid flag.
    if flags_obj.force_v2_in_keras_compile is not None:
      model.compile(
          loss='sparse_categorical_crossentropy',
          optimizer=optimizer,
          metrics=(['sparse_categorical_accuracy']
                   if flags_obj.report_accuracy_metrics else None),
          run_eagerly=flags_obj.run_eagerly,
          experimental_run_tf_function=flags_obj.force_v2_in_keras_compile)
    else:
      model.compile(
          loss='sparse_categorical_crossentropy',
          optimizer=optimizer,
          metrics=(['sparse_categorical_accuracy']
                   if flags_obj.report_accuracy_metrics else None),
          run_eagerly=flags_obj.run_eagerly)
```

@sganeshb has taken over testing.  I am not involved in TensorFlow performance going forward.

byronyi commented:
Thanks Toby @tfboyd! That’s much appreciated.

byronyi commented:
For the curious: 1d139edd6eab8f26bdde6f31f3e3709fe3cd6ed0 seems to get the performance of r1.15 back to 7600+. However after 1 single day another regression 53bdcf5 degrades the performance to <4,000 again.

tfboyd commented:
@byronyi   Wanted you to know I read your findings and nice work tracking it down.  While not good, your work shows the pain of finding old issues, they often end up layered.  I would not want you to think your work bisecting was not looked at.  It was a rough situation.  The XLA team is/was focused only on single GPU performance so they could make XLA on by default for that scenario.  They seem to be looking more holistically now.  

I am not sure there is anything we can do about it at this point other than move forward with TensorFlow 2.1.  FYI TensorFlow 2.0 100% has the XLA regression for multi-GPU.

byronyi commented:
Thanks Toby! I admit it could be difficult, but I work hard with my colleagues trying to fix these regressions for our internal 2.0 fork. 

Personally I have been looking forward to it for too long. Just can't wait there and see our users have to choose between 2.0 w/ new features and old 1.x wo/ perf regressions. We'd really like to have our cake and eat it :)

byronyi commented:
Btw, I lead in TF SIG-Networking whose goal is really just making TF runs faster on enhanced networking. Perf regressions in single-node case might seem irrelevant, but they do have a negative impact on us.

tfboyd commented:
@byronyi   Not irrelevant at all.  I have said before if you can get say 10K examples/sec on 8xV100s with FP16 + XLA but you need to scale to 32 GPUs (making up a number) to do that with FP32 then why are you focused on mulit-node and wasting money.  Figure out FP16.  There are exceptions for sure.  This is a simple mental example.  The other point being if you scale up 1 gpu and 8 GPUs (whatever your single node is) then you are scaling up everything else.

tfboyd commented:
I also agree.  It is not cool this exists in 1.x.  It came down the XLA team not wanting to focus on mulit-GPU and we stopped TF 1.x testing maybe too early.  I do not recall the final data but I stopped nightly perf tests on TF 1.x a couple months ago it was just too much for one person and there was a lack of action to resolve issues.  :-(

byronyi commented:
@anj-s I saw 53bdcf55ce180f609797746992987ce89c3543bb submitted to master again. I assume the performance regression discussed above has been fixed?

byronyi commented:
@tfboyd The aforementioned regression seems to be bd920a2427fd2c450fda41f4e8991345a7854c9f, identified by bisecting between 08-01 nightly and 08-04 nightly. This commit is also present in both r1.15 and r2.0. 

I am testing with model garden version 127b158a4c75626602f92e75582b124d142eb64b. The nightly 1.x TF packages are built using the following setup:

OS: Debian 9.9
GCC: 6.3
Bazel: 0.26.1
CUDA: 10.0
cuDNN: 7.6.1.34
TensorRT: 5.1.2

```
RUN bazel \
      --host_jvm_args=-Xmx4g \
      --host_jvm_args=-Xms512m \
      build \
        --action_env=CUDA_TOOLKIT_PATH="/usr/local/cuda" \
        --action_env=GCC_HOST_COMPILER_PATH="/usr/bin/gcc" \
        --action_env=PYTHON_BIN_PATH="/usr/bin/python3" \
        --action_env=PYTHON_LIB_PATH="/usr/local/lib/python3.5/dist-packages" \
        --action_env=TF_CONFIGURE_IOS="0" \
        --action_env=TF_CUDA_COMPUTE_CAPABILITIES="6.1,7.0,7.5" \
        --config=cuda \
        --config=numa \
        --config=tensorrt \
        --copt=-Wno-sign-compare \
        --copt=-march=ivybridge \
        --define=with_default_optimizations=true \
        --define=with_xla_support=true \
        --host_copt=-march=ivybridge \
        --python_path="/usr/bin/python3" \
        //tensorflow/tools/pip_package:build_pip_package
```

I will test by reverting it on the release branches. In the meanwhile, I'd like to ping @rachellim and see if she got any quick fixes for cherry-picking a fix instead of reverting the violating commit.

byronyi commented:
> @anj-s I saw [53bdcf5](https://github.com/tensorflow/tensorflow/commit/53bdcf55ce180f609797746992987ce89c3543bb) submitted to master again. I assume the performance regression discussed above has been fixed?

I checked that the real bug is fixed by cc9938e891ea387939c85238c9316f0c0dbb6e1a. Thanks!

rachellim commented:
Does https://github.com/tensorflow/tensorflow/commit/96a407a9186082045e368a680cc3e8af15d85d00 fix the performance issue? If so, maybe we can cherrypick it into the release?

tfboyd commented:
@rachellim Assuming it does.  It will not get into TF 2.0 as we are at the end after many weeks.  XLA is an experimental feature and doing a cherry-pick for 1.15 (same story with 2.0) may not seem risky is likely not justified form a risk reward stand point.  All of that negativity said, you can 100% petition the release owner.  I am an XLA fan for additional context.

byronyi commented:
@rachellim Cherry-Picking 96a407a9186082045e368a680cc3e8af15d85d00 along with 6d8f05acd72df61e5f4e5b4c72837b7caed3e942 does seem to fix the performance regression. Ping @jsimsal; any possibility we get these fixes into 2.0.1?

rachellim commented:
We'll cherry pick this into r1.15. As for 2.x, I'll let @goldiegadde comment further.

rachellim commented:
Note that this might not be an XLA specific issue; https://github.com/tensorflow/tensorflow/commit/96a407a9186082045e368a680cc3e8af15d85d00 affects performance & behavior of tf.data datasets with rebatching (i.e. distribution strategies).
